core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -93,53 +113,64 @@
-                             new ContentPreviewResponse(resDuplicator.duplicateStream(), contentPreviewLength))
-                .handle(voidFunction((backoffOpt, unused1) -> {
-                    if (backoffOpt.isPresent()) {
-                        final long millisAfter = useRetryAfter ?
-                                                 getRetryAfterMillis(resDuplicator.duplicateStream()) : -1;
-                        resDuplicator.close();
-                        retry0(ctx, backoffOpt.get(), rootReqDuplicator, action, deferredRes, millisAfter);
-                    } else {
-                        final HttpResponse contentPreviewResponse = new ContentPreviewResponse(
-                                resDuplicator.duplicateStream(), contentPreviewLength);
-                        contentPreviewResponse.aggregate().handle(voidFunction((unused2, cause) -> {
-                            if (cause != null && !(cause instanceof CancelledSubscriptionException)) {
-                                resDuplicator.close();
-                                final Optional<Backoff> backoffOnExceptionOpt = retryStrategy().shouldRetry(
-                                        rootReqDuplicator.duplicateStream(), cause);
-                                if (backoffOnExceptionOpt.isPresent()) {
-                                    retry0(ctx, backoffOnExceptionOpt.get(), rootReqDuplicator, action,
-                                           deferredRes, -1);
-                                } else { // exception that is not for retry occurred
-                                    deferredRes.close(cause);
-                                    rootReqDuplicator.close();
-                                }
-                            } else { // normal response
-                                deferredRes.delegate(resDuplicator.duplicateStream(true));
-                                rootReqDuplicator.close();
-                            }
-                        }));
-                    }
-                }));
+        final ContentPreviewResponse contentPreviewResponse =
+                new ContentPreviewResponse(resDuplicator.duplicateStream(), contentPreviewLength);
+        retryStrategy().shouldRetry(rootReqDuplicator.duplicateStream(), contentPreviewResponse)
+                       .handle(voidFunction((backoffOpt, unused) -> {
+                           if (backoffOpt.isPresent()) {
+                               final long millisAfter =
+                                       useRetryAfter ? getRetryAfterMillis(resDuplicator.duplicateStream())
+                                                     : -1;
+                               resDuplicator.close();
+
+                               long nextDelay;
+                               try {
+                                   nextDelay = getNextDelay(ctx, backoffOpt.get(), millisAfter);
+                                   if (nextDelay < 0) { // exceed the number of max attempt
```

Comment:
How about throwing `RetryGiveUpException` in `getNextDelay` instead of returning `-1` so that we can remove this `if` block?

Reply:
That's a great idea~!! I will change it.